### PR TITLE
#581 descriminated-unions support

### PR DESCRIPTION
**Pull request by ShravanSunder**
added contract.name to generated contract, interface and factory
allows for type manipulation and conditional types

caused by:
- #581 

related to:
- #577 

# #581 

If typechain can generate tagged unions for the contract types and factories this would make manipulation and sorting for libraries a lot easier.

For example in the definition of the following contracts
```
export class YourContract extends BaseContract {
  contractName: 'YourContract';
  connect(signerOrProvider: Signer | Provider | string): this;
...
```
```
export class Dai17 extends BaseContract {
  contractName: 'Dai17';
  connect(signerOrProvider: Signer | Provider | string): this;
...
```

I can use union discrimination and other typescript properties to properly type contracts that are stored in a library.  Such as a contract context in react.  

Related to #579  in that it allows for easy manipulation of types

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -28,3 +28,4 @@ packages/hardhat-test/artifacts
 examples/*/yarn.lock
 yarn-error.log
 packages/test-e2e/test/*/out
+.DS_Store
```
Comment (krzkaczor) [inline .gitignore:31]: Please move this to your global git ignore filee

**Review by krzkaczor — CHANGES_REQUESTED** [review]
Interesting change. "Artificial" fields like this, will require casting via `any/unknown` but I think it's required anyway currently. Can you run `yarn test:fix` to regenerate all generated files and commit them? (https://github.com/dethcrypto/TypeChain/blob/master/CONTRIBUTING.md)

File: .gitignore (hunk 1 above)
Comment (ShravanSunder) [reply .gitignore]: done

**Comment by ShravanSunder** [comment]
> Interesting change. "Artificial" fields like this, will require casting via `any/unknown` but I think it's required anyway currently. Can you run `yarn test:fix` to regenerate all generated files and commit them? (https://github.com/dethcrypto/TypeChain/blob/master/CONTRIBUTING.md)

Sorry about the delay.

@krzkaczor  I ran yarn `test:fix` however there doesn't seem to be any files  that I changed that I could commit.

image

**Comment by ShravanSunder** [comment]
@krzkaczor  hi, are things ok with this PR to merge?

File: .gitignore [hunk 2]
```diff
@@ -27,4 +27,4 @@ packages/hardhat-test/cache
 packages/hardhat-test/artifacts
 examples/*/yarn.lock
 yarn-error.log
-packages/test-e2e/test/*/out
+packages/test-e2e/test/*/out
```
Comment (krzkaczor) [inline .gitignore:30]: can you revert this change?

**Review by krzkaczor — COMMENTED** [review]
You need to commit changed files after running `yarn test` we treat these files as snapshot tests.

**Comment by ShravanSunder** [comment]
> You need to commit changed files after running `yarn test` we treat these files as snapshot tests.

hi @krzkaczor  I just ran yarn test and pushed the changes.  I also fixed gitignore ( I think master has that line included)

**Comment by krzkaczor** [comment]
@ShravanSunder thanks for working on this. Can you add me on telegram: `krzkaczor` i wanted to dive into some details.

**Comment by MHarris021** [comment]
> Interesting change. "Artificial" fields like this, will require casting via `any/unknown` but I think it's required anyway currently. Can you run `yarn test:fix` to regenerate all generated files and commit them? (https://github.com/dethcrypto/TypeChain/blob/master/CONTRIBUTING.md)

This merge breaks the ability to use generic types for Contracts as the contractName field type is now a hard-coded string. 

Previously an ERC20 that implemented IERC20 could be used in a function that declared <T extends IERC20>, i.e... 

`export async function getTokensTotalSupplies(tokens: IERC20Metadata[]) {
  tokenNumberGuard(tokens);
  return Promise.all(tokens.map(token => token.totalSupply()));
}`

would work for any Contract that could be Typecast into an @openzeppelin/contracts/token/ERC20/extensions/IERC20Metadata.sol . Now it doesn't work.

@ShravanSunder can you help me understand what additional functionality the contractName field provides to you that you couldn't get by doing a simple typeof contract or type = ContractType1|ContractType2|ContractType3 ?

**Comment by krzkaczor** [comment]
@MHarris021 thanks for this report. 

@MHarris021 @ShravanSunder do you see any other option then hiding this feature behind a flag?

**Comment by MHarris021** [comment]
@ShravanSunder it appears that you are using the contractName field on the interface here: [https://github.com/scaffold-eth/eth-hooks/blob/next/src/models/contractContextTypes.ts](https://github.com/scaffold-eth/eth-hooks/blob/next/src/models/contractContextTypes.ts) ? Is that correct?

**Comment by MHarris021** [comment]
@krzkaczor I created a generic test case not using Typechain and adding `string` to the contractName type seemed to work fine.

`export interface Stuff {
  stuffField: string|"Stuff";
  getName():string;
}

export interface AwesomeStuff {
  stuffField: string|"AwesomeStuff";
  getName():string;
  getAwesomeName():string;
}

export function doStuff<T extends Stuff>(stuff:T):string{
  return stuff.getName();
}

class AwesomeStuffImpl implements AwesomeStuff{

  stuffField: string | "AwesomeStuff";
  private readonly name: string;

  constructor() {
    this.stuffField="AwesomeStuff";
    this.name="StuffImpl";
  }
  getAwesomeName(): string {
    return `Awesome${this.getName()}`;
  }

  getName(): string {
    return this.name;
  }

}

const stuff = new AwesomeStuffImpl();

doStuff(stuff);
` 

I'd be happy to address this if you want me to.

**Comment by ShravanSunder** [comment]
> > Interesting change. "Artificial" fields like this, will require casting via `any/unknown` but I think it's required anyway currently. Can you run `yarn test:fix` to regenerate all generated files and commit them? (https://github.com/dethcrypto/TypeChain/blob/master/CONTRIBUTING.md)
> 
> This merge breaks the ability to use generic types for Contracts as the contractName field type is now a hard-coded string.
> 
> Previously an ERC20 that implemented IERC20 could be used in a function that declared , i.e...
> 
> `export async function getTokensTotalSupplies(tokens: IERC20Metadata[]) { tokenNumberGuard(tokens); return Promise.all(tokens.map(token => token.totalSupply())); }`
> 
> would work for any Contract that could be Typecast into an @openzeppelin/contracts/token/ERC20/extensions/IERC20Metadata.sol . Now it doesn't work.
> 
> @ShravanSunder can you help me understand what additional functionality the contractName field provides to you that you couldn't get by doing a simple typeof contract or type = ContractType1|ContractType2|ContractType3 ?
@MHarris021 
The reason for this is to allow for strict union discrimination when you have a Base type.  For example
```
interface FishContract extends BaseContract {
contractName : 'Fish'
}

interface MonkeyContract extends BaseContract {
contractName : 'Monkey'
}

const func = <(unknownContract: BaseContract & {contractName: string}) => {
  if (unknownContract.contractName === 'Monkey') {
     // this contract is now typed all the types can be used
  }
}

```

**Comment by ShravanSunder** [comment]
> @krzkaczor I created a generic test case not using Typechain and adding `string` to the contractName type seemed to work fine.
> 
> ```
> export interface Stuff {
>   stuffField: string|"Stuff";
>   getName():string;
> }
> 
> export interface AwesomeStuff {
>   stuffField: string|"AwesomeStuff";
>   getName():string;
>   getAwesomeName():string;
> }
> 
> export function doStuff<T extends Stuff>(stuff:T):string{
>   return stuff.getName();
> }
> 
> class AwesomeStuffImpl implements AwesomeStuff{
> 
>   stuffField: string | "AwesomeStuff";
>   private readonly name: string;
> 
>   constructor() {
>     this.stuffField="AwesomeStuff";
>     this.name="StuffImpl";
>   }
>   getAwesomeName(): string {
>     return `Awesome${this.getName()}`;
>   }
> 
>   getName(): string {
>     return this.name;
>   }
> 
> }
> 
> const stuff = new AwesomeStuffImpl();
> 
> doStuff(stuff); 
> ```
> 
> I'd be happy to address this if you want me to.

as I mentioned in #603 this change does not work as explained in the above post

@MHarris021  @krzkaczor  we could make it an option in typechain.  I think that works for me.

**Comment by ShravanSunder** [comment]
@MHarris021  as a temporary workaround.
you can cast into IERC20Metadata:
export async function getTokensTotalSupplies(tokens: IERC20Metadata[]) { tokenNumberGuard(tokens); return Promise.all(tokens.map(token => token.totalSupply())); }

if you're calling this function.  if IERC20Metadata is based on typechain.  
https://www.typescriptlang.org/docs/handbook/utility-types.html#omittype-keys
```
getTokensTotalSupplies(tokens: Omit<IERC20Metadata, 'contractName'>[]) {
...
}
```

**Comment by Lumyo** [comment]
@ShravanSunder I tried the suggested workaround, but it doesnt work and I'm still getting erros, because there are a bunch of nested `contractName` fields which arent ommited by this (Ex : `interface.contractName`, `deployed().contractName`, `connect().contractName`, ...)

**Comment by ShravanSunder** [comment]
> @ShravanSunder I tried the suggested workaround, but it doesnt work and I'm still getting erros, because there are a bunch of nested `contractName` fields which arent ommited by this (Ex : `interface.contractName`, `deployed().contractName`, `connect().contractName`, ...)

If you share your branch in your fork I can try to take a look this weekend
